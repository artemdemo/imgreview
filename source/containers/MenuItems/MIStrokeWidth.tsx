import React from 'react';
import { connect } from 'react-redux';
import { TReduxState } from '../../reducers';
import Icon from '../../components/Icon/Icon';
import MainMenuItem from '../../components/MainMenu/MainMenuItem';
import { setStrokeWidth, TSetStrokeWidth } from '../../model/shapes/shapesActions';
import { TStateShapes } from '../../model/shapes/shapesReducer';

type Props = {
    shapes: TStateShapes;
    setStrokeWidth: TSetStrokeWidth;
    disabled: boolean;
};

class MIStrokeWidth extends React.PureComponent<Props> {
    static readonly defaultProps = {
        disabled: false,
    };

    handleSubMenuClick = (item, e) => {
        e.stopPropagation();
        const { setStrokeWidth } = this.props;
        setStrokeWidth(item.value);
    };

    render() {
        const { shapes, disabled } = this.props;
        return (
            <MainMenuItem
                subMenu={[
                    {
                        text: '2px',
                        value: 2,
                        selected: shapes.strokeWidth === 2,
                        onClick: this.handleSubMenuClick,
                    },
                    {
                        text: '3px',
                        value: 3,
                        selected: shapes.strokeWidth === 3,
                        onClick: this.handleSubMenuClick,
                    },
                    {
                        text: '5px',
                        value: 5,
                        selected: shapes.strokeWidth === 5,
                        onClick: this.handleSubMenuClick,
                    },
                    {
                        text: '8px',
                        value: 8,
                        selected: shapes.strokeWidth === 8,
                        onClick: this.handleSubMenuClick,
                    },
                ]}
                disabled={disabled}
            >
                <Icon
                    name='pencil'
                    title='Size'
                />
            </MainMenuItem>
        );
    }
}

export default connect(
    (state: TReduxState) => ({
        shapes: state.shapes,
    }),
    {
        setStrokeWidth,
    }
)(MIStrokeWidth);
