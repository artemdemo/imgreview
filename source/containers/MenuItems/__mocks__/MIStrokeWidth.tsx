import React from 'react';

const MIStrokeWidth = props => (
    <div data-mock='MIStrokeWidth'>
        {props.children}
    </div>
);

export default MIStrokeWidth;
