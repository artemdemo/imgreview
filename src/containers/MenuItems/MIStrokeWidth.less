.MIStrokeWidth__Content {
  display: inline-block;
}
