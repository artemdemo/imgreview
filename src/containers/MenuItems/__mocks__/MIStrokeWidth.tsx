import React from 'react';

const MIStrokeWidth = props => (
    <div data-mock='MIStrokeWidth'>
        {JSON.stringify(props, null, 2)}
    </div>
);

export default MIStrokeWidth;
