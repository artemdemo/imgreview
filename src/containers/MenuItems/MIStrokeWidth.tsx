import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';
import { TReduxState } from '../../reducers';
import { TopMenuItem } from '../../components/TopMenu/TopMenuItem';
import ModalClickOutside from '../../components/Modal/ModalClickOutside';
import { setStrokeWidth, toggleSubmenu } from '../../model/menu/menuActions';
import { TStateMenu } from '../../model/menu/menuReducer';
import * as api from '../../../srcCanvas/api';
import * as gaService from '../../services/ganalytics';
import { EIcon, ImgIcon } from './ImgIcon/ImgIcon';
import * as canvasApi from '../../../srcCanvas/api';
import IShape from '../../../srcCanvas/Shape/IShape';
import store from '../../store';
import './MIStrokeWidth.less';

const handleDragStarted = (shape: IShape) => {
  if (_.isFunction(shape.getStrokeWidth)) {
    store.dispatch(setStrokeWidth(shape.getStrokeWidth()));
  }
};

canvasApi.shapeClicked.on(handleDragStarted);
canvasApi.shapeDragStarted.on(handleDragStarted);

const STROKE_WIDTH = 'STROKE_WIDTH';

type Props = {
  disabled: boolean;
  show: boolean;
};

export const MIStrokeWidth: React.FC<Props> = (props) => {
  const menu = useSelector<TReduxState, TStateMenu>((state) => state.menu);
  const dispatch = useDispatch();
  const { disabled, show } = props;

  const handleSubMenuClick = (item: any) => {
    dispatch(setStrokeWidth(item.value));
    api.setStrokeWidthToActiveShape(item.value);

    gaService.sendEvent({
      eventCategory: gaService.EEventCategories.MenuClick,
      eventAction: gaService.EEventActions.ChangeStrokeWidth,
    });
  };

  const createSubmenuItem = (value: number) => {
    return {
      text: `${value}px`,
      value,
      selected: menu.strokeWidth === value,
      onClick: handleSubMenuClick,
    };
  };

  const handleMenuClick = () => {
    dispatch(toggleSubmenu(menu.openSubmenu === '' ? STROKE_WIDTH : ''));
  };

  const handleClickOutside = () => {
    if (menu.openSubmenu === STROKE_WIDTH) {
      // There is weird bug with events propagation,
      // if I'm not wrapping these events dispatching.
      // (User can't add Arrow shape to the scene)
      requestAnimationFrame(() => {
        dispatch(toggleSubmenu(''));
      });
    }
  };

  const values = [2, 3, 5, 6, 7, 8, 10];
  return (
    <ModalClickOutside onClickOutside={handleClickOutside}>
      <TopMenuItem
        subMenu={values.map(createSubmenuItem)}
        open={menu.openSubmenu === STROKE_WIDTH}
        disabled={disabled}
        show={show}
        onClick={handleMenuClick}
      >
        <span className="MIStrokeWidth__Content">{menu.strokeWidth}px</span>
        <ImgIcon icon={EIcon.strokeWidth} />
      </TopMenuItem>
    </ModalClickOutside>
  );
};
