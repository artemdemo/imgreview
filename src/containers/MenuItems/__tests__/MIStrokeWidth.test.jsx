import React from 'react';
import { mount } from 'enzyme';
import MIStrokeWidth from '../MIStrokeWidth';
import renderer from 'react-test-renderer';

jest.mock('react-redux');
jest.mock('../../../components/TopMenu/TopMenuItem');
jest.mock('../../../../srcCanvas/api');
jest.mock('../../../services/ganalytics');

const eventMock = {
    stopPropagation: jest.fn(),
};

const apiMock = require('../../../../srcCanvas/api');

describe('MIStrokeWidth', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should not render', () => {
        const tree = renderer.create(
            <MIStrokeWidth
                menu={{ strokeWidth: 5 }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render', () => {
        const tree = renderer.create(
            <MIStrokeWidth
                menu={{ strokeWidth: 5 }}
                show
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render disabled', () => {
        const tree = renderer.create(
            <MIStrokeWidth
                menu={{ strokeWidth: 5 }}
                show
                disabled
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should handle menu click', () => {
        const toggleSubmenu = jest.fn();
        const wrapper = mount(
            <MIStrokeWidth
                menu={{ strokeWidth: 5 }}
                toggleSubmenu={toggleSubmenu}
            />
        );

        const instance = wrapper.instance();

        instance.handleMenuClick(eventMock);

        expect(eventMock.stopPropagation).toBeCalledWith();
        expect(toggleSubmenu).toBeCalledWith('');
    });

    it('should handle sub-menu click', () => {
        const setStrokeWidthMock = jest.fn();
        const wrapper = mount(
            <MIStrokeWidth
                menu={{ strokeWidth: 5 }}
                setStrokeWidth={setStrokeWidthMock}
            />
        );

        const item = {
            value: 'some value',
        };

        const instance = wrapper.instance();

        instance.handleSubMenuClick(item, eventMock);

        expect(setStrokeWidthMock).toBeCalledWith(item.value);
        expect(apiMock.setStrokeWidthToActiveShape).toBeCalledWith(item.value);
    });
});
