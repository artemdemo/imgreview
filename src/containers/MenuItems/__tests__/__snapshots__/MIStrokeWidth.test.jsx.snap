// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MIStrokeWidth should not render 1`] = `null`;

exports[`MIStrokeWidth should render 1`] = `
<div
  data-mock="TopMenuItem"
  onClick={[Function]}
>
  <div
    data-mock="Icon"
  />
</div>
`;

exports[`MIStrokeWidth should render disabled 1`] = `
<div
  data-mock="TopMenuItem"
  onClick={[Function]}
>
  <div
    data-mock="Icon"
  />
</div>
`;
